### ISSUE TITLE
Support running benchmarks without creating benchmark jar?

### ISSUE THREAD

**Issue by cursive-ide** [melix/jmh-gradle-plugin#69]
I'm trying to benchmark a class from my project. My project is part of an IntelliJ plugin, so it needs the whole IntelliJ framework on the classpath even though it needs only a couple of classes from it. When trying to run my benchmark, I get:

> Execution failed for task ':jmhJar'.
> archive contains more than 65535 entries.

I guess this is because it's trying to put the whole IntelliJ framework into the benchmark jar. Is it possible to run JMH via this plugin without creating the jar?

**Comment by aalmiray** [comment]
1. does it really need the whole IntelliJ classes in the classpath in order to run?
2. did you apply the shadow plugin?

@vyazelenko methinks we may need to provide an exclusion block for the JmhJar task that works with both plain and shadow versions.

**Comment by cursive-ide** [comment]
1. No, it doesn't, but I'd have to manually filter out the ones it doesn't need, which is annoying. There's no good way to choose the ones it needs.
2. No, but the IntelliJ community edition is like 235MB, I don't want to have to create a jar for that.

**Comment by cursive-ide** [comment]
BTW as a workaround, I've been trying this:
```
task lexerBench(type: JavaExec, dependsOn: [jmhClasses]) {
  main = "benchmark.FlexLexer"
  classpath = configurations.jmh
}
```
However this doesn't work, because my benchmark class does not appear in the build directory. Is this because it's using the bytecode generator and not the annotation processor?

**Comment by aalmiray** [comment]
I think there may be a misunderstanding in the usage of shadow for 2.; the jmh plugin creates a fat jar of your production classes, alongside JMH and project dependencies. This taks is performed using a regular Jar gradle task. However, if the `shadow` plugin is applied then a different task (ShadowJar) is used. This is because `shadow` has better support for merging duplicate files.

Regardless of which approach is used, the jmhJar task will include everything, which causes you trouble. It looks to me that the jmh plugin should allow developers to define a set of includes/excludes (using path regexes) to determine which classes should be added to the fat jar.

**Comment by cursive-ide** [comment]
I think that would help, but it would still be tedious in my case - I would have to manually identify the IntelliJ classes used by my benchmark, along with the classes transitively required by them, and manually include them. A better solution for my case, at least, would be to support running the benchmarks via the `main` method.

**Comment by vyazelenko** [comment]
@cursive-ide you need to enable `zip64` in order for jar task to succeed, i.e.:
```gradle
jmh {
    zip64 = true
}
```
Btw, packaging benchmarks into fat jar is the intended way advocated by JMH itself, see [JMH project documentation](http://openjdk.java.net/projects/code-tools/jmh/).

**Comment by vyazelenko** [comment]
@cursive-ide FYI, since 0.3.0 release benchmarks are generated by `jmhCompileGeneratedClasses` task and compiled classes can be found inside `$project.buildDir/jmh-generated-classes` directory.

**Comment by austinzh** [comment]
@vyazelenko I agree for real benchmark it's better using fat jar, but a lot of time we need to test the benchmark code as well.
In our use case, we need to build a fat jar around 1G to benchmark. Surely we can argue that we can reduce the size of the jar if we choice what should put into it. But since we already has all the artifact build. Would it be nice to have a way to run from classpath directly as a option so that the jmhJar don't need to build everytime?

**Comment by austinzh** [comment]
@melix I have try to run without this plugin as [this article](https://arbitrary-but-fixed.net/java/algorithms/2019/09/29/jmh-gradle-config.html) suggested  and always end out fail on `Unable to find the resource: /META-INF/BenchmarkList`

**Comment by austinzh** [comment]
For folks that want to run without build jar. Here is my build.gradle.kts
```
tasks.register<JavaExec>("runJmh") {
    group = "jmh"
    description = "Run JMH benchmarks"
    dependsOn("jmhCompileGeneratedClasses")

    val jmhGeneratedClassesDir = layout.buildDirectory.dir("jmh-generated-classes").get().asFile
    val jmhGeneratedResourcesDir = layout.buildDirectory.dir("jmh-generated-resources").get().asFile
    classpath = sourceSets["jmh"].runtimeClasspath + sourceSets["main"].runtimeClasspath + files(jmhGeneratedClassesDir) + files(jmhGeneratedResourcesDir)
    mainClass.set("org.openjdk.jmh.Main")
}
```
